geniza/common/middleware.py

```diff
@@ -0,0 +1,46 @@
+from django.conf import settings
+from django.http import HttpRequest, HttpResponse, HttpResponseRedirect
+from django.urls import is_valid_path
+from django.utils import translation
+from django.utils.deprecation import MiddlewareMixin
+
+
+class PublicLocaleMiddleware(MiddlewareMixin):
+    """Middleware to redirect anonymous users attempting to access locales that are not in
+    the list of public site languages."""
+
+    response_redirect_class = HttpResponseRedirect
+
+    def process_request(self, request: HttpRequest):
+        public_site_languages = getattr(settings, "PUBLIC_SITE_LANGUAGES", [])
+        if public_site_languages and not request.user.is_authenticated:
+            language_from_path = translation.get_language_from_path(request.path_info)
+            if language_from_path and language_from_path not in public_site_languages:
+                # If the path language isn't in the public site languages list,
+                # update cookies and the request object to use the default language
+                request.COOKIES[settings.LANGUAGE_COOKIE_NAME] = settings.LANGUAGE_CODE
+                translation.activate(settings.LANGUAGE_CODE)
+                request.LANGUAGE_CODE = translation.get_language()
```

Comment:
surprised we have to do the explicit change, is there no way to use the `set_language` redirect view? https://docs.djangoproject.com/en/4.0/topics/i18n/translation/#the-set-language-redirect-view

Reply:
Good call, I forgot about `set_language`! Done in d9a7304 👍